services/galley/src/Galley/API/Query.hs

```diff
@@ -143,6 +149,21 @@ getConversations user mids mstart msize = do
       | Data.isConvDeleted c = Data.deleteConversation (Data.convId c) >> pure False
       | otherwise = pure True
 
+iterateConversations :: forall a. UserId -> Range 1 500 Int32 -> ([Data.Conversation] -> Galley a) -> Galley [a]
+iterateConversations uid pageSize handleConvs = catMaybes <$> go Nothing
+  where
+    go :: Maybe ConvId -> Galley [Maybe a]
+    go mbConv = do
+      convResult <- getConversationsInternal uid Nothing mbConv (Just pageSize)
+      resultHead <- Just <$> handleConvs (convList convResult)
+      resultTail <- case convList convResult of
+        (conv : rest) ->
+          if convHasMore convResult
+            then go (Just (maximum (Data.convId <$> (conv : rest))))
```

Comment:
I'm still not sure I understand this.  Either the convIds come back from cassandra in increasing order, in which case `maximum` should be `last`.  Or they don't, in which case `maximum` won't work, because it may miss arbitrarily many entries.  Please enlighten me!

Reply:
I agree that `last` would be equivalent. I wasn't sure if cassandra returns them ordered or not, so I didn't want to take any chances.
Why would `maximum` miss anything? Unless I'm mistaking `maximum` always returns the biggest element of a non-empty data structure, no matter if the elements are ordered or not. The default implementation of `Foldable` class is
```haskell
maximum :: forall a . Ord a => t a -> a
maximum = fromMaybe (errorWithoutStackTrace "maximum: empty structure") .
 getMax . foldMap (Max #. (Just :: a -> Maybe a))
```

Reply:
ah, good point.  let's keep it as is then.